tests/Functional/ResultCacheTest.php

```diff
@@ -225,6 +225,64 @@ public function testFetchAllSavesCache(): void
         self::assertCount(1, $layerCache->fetch('testcachekey'));
     }
 
+    public function testCacheQueriedOnlyOnceForCacheMiss(): void
+    {
+        $layerCache = $this->createMock(ArrayCache::class);
+        $layerCache->expects(self::once())
+            ->method('fetch')
+            ->willReturn(false);
+
+        $layerCache->expects(self::once())
+            ->method('save');
+
+        $result = $this->connection->executeQuery(
+            'SELECT * FROM caching WHERE test_int > 500',
+            [],
+            [],
+            new QueryCacheProfile(0, 'testcachekey', $layerCache)
+        );
+        $result->fetchAllAssociative();
+    }
+
+    public function testDifferentQueriesMayBeSavedToOneCacheKey(): void
```

Comment:
This test passes even without the fix. Is it really needed?

Reply:
It shows how many queries may be stored in one cache key. This logic wasn't explicitly shown in other tests and you weren't aware of it, so I think it's better to leave this test. However, it might be taken out to another PR, as it doesn't directly relates to this one.